truss/contexts/image_builder/cache_warmer.py

```diff
@@ -48,106 +66,141 @@ def split_path(path, prefix="gs://"):
+        if not blob.exists(self.client):
+            raise RuntimeError(f"File not found on GCS bucket: {blob.name}")
 
+        if self.is_private:
             url = blob.generate_signed_url(
                 version="v4",
                 expiration=datetime.timedelta(minutes=15),
                 method="GET",
             )
-            try:
-                proc = _download_from_url_using_b10cp(_b10cp_path(), url, dst_file)
-                proc.wait()
-            except Exception as e:
-                raise RuntimeError(f"Failure downloading file from GCS: {e}")
-        elif prefix == "s3://":
+        else:
+            base_url = "https://storage.googleapis.com"
+            url = f"{base_url}/{self.bucket_name}/{blob.name}"
+        try:
+            proc = _download_from_url_using_b10cp(_b10cp_path(), url, dst_file)
+            proc.wait()
+        except Exception as e:
+            raise RuntimeError(f"Failure downloading file from GCS: {e}")
+
+
+class S3File(RepositoryFile):
+    def connect(self, key_file="/app/data/s3_credentials.json"):
+        self.bucket_name, _ = split_path(repo_name, prefix="s3://")
+
+        if os.path.exists(key_file):
             (
                 AWS_ACCESS_KEY_ID,
                 AWS_SECRET_ACCESS_KEY,
                 AWS_REGION,
             ) = parse_s3_service_account_file(key_file)
-            client = boto3.client(
+            self.client = boto3.client(
                 "s3",
                 aws_access_key_id=AWS_ACCESS_KEY_ID,
                 aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
                 region_name=AWS_REGION,
                 config=Config(signature_version="s3v4"),
             )
-            bucket_name, _ = split_path(bucket_name, prefix="s3://")
-
-            dst_file = Path(f"{cache_dir}/{file_name}")
-            if not dst_file.parent.exists():
-                dst_file.parent.mkdir(parents=True)
-
-            try:
-                url = client.generate_presigned_url(
-                    "get_object",
-                    Params={"Bucket": bucket_name, "Key": file_name},
-                    ExpiresIn=3600,
-                )
-            except Exception:
-                raise RuntimeError(f"File not found on S3 bucket: {file_name}")
-
-            try:
-                proc = _download_from_url_using_b10cp(_b10cp_path(), url, dst_file)
-                proc.wait()
-            except Exception as e:
-                raise RuntimeError(f"Failure downloading file from S3: {e}")
-    else:
-        secret_path = Path("/etc/secrets/hf-access-token")
-        secret = secret_path.read_text().strip() if secret_path.exists() else None
+        else:
+            self.client = boto3.client("s3", config=Config(signature_version=UNSIGNED))
+
+    def download_to_cache(self):
+        cache_dir = Path(f"/app/model_cache/{self.bucket_name}")
+        cache_dir.mkdir(parents=True, exist_ok=True)
+
+        dst_file = Path(f"{cache_dir}/{self.file_name}")
```

Comment:
I think `dst_file = cache_dir / self.file_name` or `dst_file = cache_dir / Path(self.file_name)` should work here